benchmark_runner/jupyterlab/templates/summary_report/summary_report_operations.py

```diff
@@ -0,0 +1,353 @@
+
+import logging
+import pandas as pd
+from functools import reduce
+from datetime import datetime
+from typeguard import typechecked
+from benchmark_runner.common.elasticsearch.elasticsearch_operations import ElasticSearchOperations
+
+# Configure the logger
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+class SummaryReportOperations:
+    """ This class contains summary report operations """
+    GB = 1_000_000_000
+    MILLISECONDS = 1000
+
+    def __init__(self, elasticsearch: ElasticSearchOperations):
+        self.elasticsearch = elasticsearch
+        self.geometric_mean_data = None
+        self.geometric_mean_data_throughput = None
+        self.geometric_mean_data_iops = None
+        self.geometric_mean_data_latency = None
+        self.bootstorm_df = None
+        self.median_indices = None
+        self.geometric_mean_df = None
+        self.operator_versions = None
+
+    def get_operator_versions(self):
+        """
+        This method returns operator versions
+        @return:
+        """
+        return self.operator_versions
+
+    @typechecked
+    def get_workload_data(self, start_datetime: datetime, end_datetime: datetime, index: str ='ci-status'):
+        """
+        This method returns workload data from elasticsearch
+        @return:
+        """
+
+        return self.elasticsearch.get_index_data_between_dates(index=index, start_datetime=start_datetime, end_datetime=end_datetime)
+
+    @typechecked
+    def filter_data(self, entries: list, compared_ocp_versions: tuple, kind: str, filter_field1: dict = None,
+                    filter_field2: dict = None):
+        """
+        This method filters elastic search data by filter fields
+        @param entries:
+        @param compared_ocp_versions:
+        @param kind:
+        @param filter_field1:
+        @param filter_field2:
+        @return:
+        """
+        filter_data = [
+            entry.get('_source') for entry in entries
+            if (
+                    entry.get('_source').get('ocp_version') in compared_ocp_versions and
+                    entry.get('_source').get('kind') == kind and
+                    (filter_field1 is None or entry.get('_source').get(
+                        list(filter_field1.keys())[0]) == list(filter_field1.values())[0]) and
+                    (filter_field2 is None or (list(filter_field2.keys())[0] in entry.get('_source') and
+                                               entry.get('_source').get(list(filter_field2.keys())[0]) == list(filter_field2.values())[0]))
+            )
+        ]
+        return pd.DataFrame(filter_data)
+
+    @staticmethod
+    def __geometric_mean(values: list):
+        """
+        This method calculates geometric means for input values
+        @param values:
+        @return: values geometric means
+        """
+        product = reduce(lambda x, y: x * y, values)
+        mean = pow(product, 1 / len(values))
```

Comment:
If you have a lot of values, this can overflow even double precision floats.  Do it by averaging the log of the values, and then exponentiating the result.

Reply:
I am filtering the data before it reach to this part, meaning it will calc geometric mean only up 2 ocp versions

Reply:
Yeah, OK, but it's a good idea to do this correctly.

Reply:
What do you recommend ?